source/extensions/filters/http/on_demand/on_demand_update.cc

```diff
@@ -0,0 +1,63 @@
+#include "extensions/filters/http/on_demand/on_demand_update.h"
+
+#include "common/common/assert.h"
+#include "common/common/enum_to_int.h"
+#include "common/http/codes.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace OnDemand {
+
+void OnDemandRouteUpdate::requestRouteConfigUpdate() {
+  if (callbacks_->route() != nullptr || !callbacks_->canRequestRouteConfigUpdate()) {
+    filter_return_ = FilterReturn::ContinueDecoding;
+  } else {
+    callbacks_->requestRouteConfigUpdate([this]() -> void { onRouteConfigUpdateCompletion(); });
+    filter_return_ = FilterReturn::StopDecoding;
+  }
+}
+
+Http::FilterHeadersStatus OnDemandRouteUpdate::decodeHeaders(Http::HeaderMap&, bool) {
+  requestRouteConfigUpdate();
+  return filter_return_ == FilterReturn::StopDecoding ? Http::FilterHeadersStatus::StopIteration
+                                                      : Http::FilterHeadersStatus::Continue;
+}
+
+Http::FilterDataStatus OnDemandRouteUpdate::decodeData(Buffer::Instance&, bool) {
+  return filter_return_ == FilterReturn::StopDecoding
+             ? Http::FilterDataStatus::StopIterationAndWatermark
+             : Http::FilterDataStatus::Continue;
+}
+
+Http::FilterTrailersStatus OnDemandRouteUpdate::decodeTrailers(Http::HeaderMap&) {
+  return Http::FilterTrailersStatus::Continue;
+}
+
+void OnDemandRouteUpdate::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callbacks) {
+  callbacks_ = &callbacks;
+}
+
+// This is the callback which is called when an update requested in requestRouteConfigUpdate()
+// has been propagated to workers, at which point the request processing is restarted from the
+// beginning.
+void OnDemandRouteUpdate::onRouteConfigUpdateCompletion() {
+  filter_return_ = FilterReturn::ContinueDecoding;
+
+  if (callbacks_->canResolveRouteAfterConfigUpdate() && // route can be resolved after an on-demand
+                                                        // VHDS update
+      !callbacks_->decodingBuffer() &&                  // Redirects with body not yet supported.
+      callbacks_->recreateStream()) {
```

Comment:
could you help me to understand why we need to recreatestream here?

Reply:
I'm doing an internal redirect here to restart the current ActiveConnection and pick up the updated vhosts/routing table.

Reply:
My question is why not just clearRouteCache() + refreshCachedRoute?

Reply:
I don't think this would work, as ```snapped_route_config_``` would still be holding a pointer to the pre-update route configuration. Additionally, filter chain creation is route-dependent, a side-effect of route being resolvable now is that the filter chain could be different now too.

Reply:
hmm, right, we should not have two different versions of snapped_route_config_ in one httprequest's process, even if we can do so. 
OTOH, could you refresh me where is the route-dependent part of a filter-chain creation?